dbms/src/Storages/Transaction/LearnerRead.cpp

```diff
@@ -55,18 +55,18 @@ struct UnavailableRegions
 
     bool empty() const { return size() == 0; }
 
-    void setRegionLock(RegionID region_id_, LockInfoPtr && region_lock_)
+    void addRegionLock(RegionID region_id_, LockInfoPtr && region_lock_)
     {
-        region_lock = std::pair(region_id_, std::move(region_lock_));
+        region_locks.emplace_back(region_id_, std::move(region_lock_));
         doAdd(region_id_);
     }
 
     void tryThrowRegionException(bool batch_cop)
     {
         // For batch-cop request, all unavailable regions, include the ones with lock exception, should be collected and retry next round.
         // For normal cop request, which only contains one region, LockException should be thrown directly and let upper layer(like client-c, tidb, tispark) handle it.
-        if (!batch_cop && region_lock)
-            throw LockException(region_lock->first, std::move(region_lock->second));
+        if (!batch_cop && !region_locks.empty())
+            throw LockException(std::move(region_locks));
```

Comment:
When disagg read, we will call `doLearnerRead(... /*for_batch_cop=*/true, ...)`. So I think it will not run into this if branch and no LockException is thrown under disagg read now.
This is sufficient for batch cop or MPP, since they handle `RegionException` and `LockException` in the same way. But disagg read need to acquire the LockInfo in `FlashService::EstablishDisaggTask` and send it back to RN. So I think the changes now is not sufficient.

https://github.com/pingcap/tiflash/blob/c83a185453b4b5a3cb8ad43307ee761ae725fada/dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp#L558-L566

P.S. Can we add a unit test to ensure the behavior that Region lock is thrown correctly to the upper level?

Reply:
How about changing to batch cop first, and then I update this PR? Currently it is hard for me to check whether the change "will" work under batch cop.

Reply:
now it is already running with "batchcop =true"

Reply:
if (dag_context.isBatchCop() || dag_context.isMPPTask() || dag_context.is_disaggregated_task)

Reply:
https://github.com/pingcap/tiflash/blob/ea748e8c4b60902f6d18d642cd279708d26afbf3/dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp#L379-L382